src/Common/src/System/Numerics/Hashing/Hash.cs

```diff
@@ -0,0 +1,103 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace System.Numerics.Hashing
+{
+    internal static class Hash
+    {
+        public static int Combine(int h1, int h2)
+        {
+            // The jit optimizes this to use the ROL instruction on x86
+            // Related GitHub pull request: dotnet/coreclr#1830
+            uint shift5 = ((uint)h1 << 5) | ((uint)h1 >> 27);
+            return ((int)shift5 + h1) ^ h2;
```

Comment:
If @VSadov and @jcouv are happy with the revised hash function, then that could be changed in-place in ValueTuple.  Separating out all of these methods into their own internal common class just doesn't seem worth it to me; if all of them are there, then any code which wants to include just one will need to compile in them all (which affects IL size, makes it more difficult to understand what code is dead, etc.), and if only some are here, then ValueTuple will be calling to the common class for some of the combinations and doing some itself, which is in and of itself confusing.  If at some point we wanted to expose a public class for this, then it may make sense, though even then it's not clear to me we'd want all of these overloads.  In the meantime, I'd be ok with having the int,int hash combining function itself in a common class, so that various types can use the same hash function, but everything else seems superfluous to me.